### PR TITLE
Add override support to auto-generator

### LINKED ISSUE #19: Add ability to override custom types in auto-generator
It would be great if the user could specify overrides for any complex type they desire. For example, say `Customer` is a type that has a field of type `Customer list` (sub-customers of the current customer). This is recursive and (currently) needs a specifically designed generator to avoid stack overflow. Say then that I want to generate an `Order` type which, among other things, contains a `Customer`, then I'd want to just say `auto<Order>(...)` and somehow specify that I want `Customer` objects to be generated using my specific generator, and otherwise the defaults are fine.

I'm not sure how to do this though, or if it is easy/feasible/possible.

### PR DESCRIPTION
**Pull request by cmeeren**
Closes #19 

/cc @moodmosaic @TysonMN 

The tests demonstrate how to use it. The readme will be updated too; for now, let's just settle the API and implementation.

I suggest you hide whitespace changes when viewing the diff, since most of `auto` has been indented.

Todo:

* [ ] Update readme

### CODE REVIEW

**Comment by TysonMN** [comment]
@dharmaturtle, want to try this out on your code?

**Comment by cmeeren** [comment]
An advanced feature could be to somehow make the overrides hierarchical. E.g., when generating `MyType`, override the `int` and `string` generators, otherwise use the default auto-generators.

Unless someone really needs it (or someone knows of a very simple way to implement it), I won't prioritize that now.

**Comment by dharmaturtle** [comment]
I made some small edits, changing the type from Map to List. Thoughts?

https://github.com/hedgehogqa/fsharp-hedgehog-experimental/compare/auto-overrides...dharmaturtle:auto-overrides

I'm not entirely happy with line 395, but for the life of me I can't get Reflection to return the "real" generic type and not `System.Object`.

---

On a different note, this PR solves the problem I was having with NodaTime!

image

**Comment by cmeeren** [comment]
@dharmaturtle Why the change to `list`? I can't see any other effect of this than poorer lookup performance, as well as being more error-prone since it will fail if more than one generator is added for any given type (whereas with the `Map` approach the last one to be added for any given type will be the one that's used, which allows you to e.g. create a default config with all your domain types and still override certain types in some tests.)

File: src/Hedgehog.Experimental/Gen.fs [hunk 1]
```diff
@@ -24,8 +24,16 @@ type AutoGenConfig = {
...
   SeqRange : Range<int>
   RecursionDepth: int
+  Overrides: Map<string, Gen<obj>>
 }
 
+
+module AutoGenConfig =
+
+  let addOverride (gen: Gen<'a>) config =
+    { config with Overrides = config.Overrides.Add (typeof<'a>.FullName, Gen.map box gen) }
```
Comment (TysonMN) [inline src/Hedgehog.Experimental/Gen.fs:34]: What about using `typeof<'a>` as the key instead of `typeof<'a>.FullName`?

Comment (cmeeren) [reply src/Hedgehog.Experimental/Gen.fs:34]: That was what I first tried, but `Type` doesn't support the `comparison` constraint which is required for `Map` keys.

**Comment by dharmaturtle** [comment]
I thought `list` was preferable because it follows the spirit of "makes illegal states unrepresentable". E.g. with `list`, you can't do the following:

```
{ GenX.defaults with Overrides = Map.empty.Add("string", Gen.int) }
```

(Not literal, as you'd need to `box` up the Gen and give it a `Range`, but I'm just trying to illustrate a point.)

Re: perf: that shouldn't be a concern for this level of scale. I prefer correctness over perf for now. If it really does become an issue we can always encapsulate an internal `map` or whatever.

Re: multiple gen: I flipped around the args so the new gen comes first:

```
{ config with Overrides = Gen.map box gen :: config.Overrides }
```

Later, I do `Seq.tryHead` so it'll grab the first/latest gen, preserving the `Map`-style behavior.

**Comment by cmeeren** [comment]
> I thought `list` was preferable because it follows the spirit of "makes illegal states unrepresentable". E.g. with `list`, you can't do the following:
> 
> ```
> { GenX.defaults with Overrides = Map.empty.Add("string", Gen.int) }
> ```

The intention is for users to primarily use `AutoGenConfig.addOverride`. Ideally I'd like to have `Overrides` be internal, but that's not possible without making the whole record internal, which is a breaking change and not something I'd want anyway. (Though one could argue for ditching the `{ GenX.defaults with ... }` syntax entirely and creating functions in the `AutoGenConfig` module for setting everything.)

So what I'm thinking is that I just document that the `Overrides` map are boxed gens keyed by the `System.Type.FullName` of the original type. I think that this is a simple and easy-to-understand contract that should be adequate for Hedgehog users (who are, after all, likely to be advanced devs in the first place) *if* they at all want to use `Overrides` directly instead of `addOverride`.

> Re: perf: that shouldn't be a concern for this level of scale.

The generators are effectively run in tight loops, so I certainly wouldn't dismiss it out of hand.

> I'm not entirely happy with line 395, but for the life of me I can't get Reflection to return the "real" generic type and not `System.Object`.

Yes, I'm also not happy about this line, since it has to actually generate a value of potentially all the generators in the list (particularly in the case that it needs to generate a value that is not overridden) in order to reflect on it so that it knows if it should use the generator. See my previous remarks regarding performance.

**Comment by TysonMN** [comment]
> The intention is for users to primarily use `AutoGenConfig.addOverride`. Ideally I'd like to have `Overrides` be internal, but that's not possible without making the whole record internal, which is a breaking change and not something I'd want anyway.

Instead of `Overrides` having type `Map<string, Gen<obj>>`, it could be of type (say) `AutoOverrides` and that could be a record with a private constructor.  Then the only way to add an override would be via `AutoGenConfig.addOverride`.  Then from the user's perspective, there is no way to represent illegal state.

**Comment by cmeeren** [comment]
Thanks @TysonMN, excellent tip. Updated the PR now.

**Comment by cmeeren** [comment]
Come to think of it, is there any reason now to have members for all the default generators on `AutoGenConfig`? Couldn't they just be defaults added to the map?

**Comment by TysonMN** [comment]
I also like the second one better.  I am confused though about what the breaking change is.  Can you elaborate?

**Comment by cmeeren** [comment]
I am talking about removing all the primitive `Byte`, `Int16` etc. members of `AutoGenConfig` and not even using them internally, instead relying completely on the `Overrides` member (which should be named `Generators` or something like that). Simply put, this would be the new definition (keeping the `Overrides` name for now):

```f#
[<CLIMutable; Struct>]
type AutoGenConfig = {
  SeqRange : Range<int>
  RecursionDepth: int
  Overrides: AutoOverrides
}
```

This is a breaking change, since users can't do this anymore:

```f#
{ GenX.defaults with Foo = myGenerator }
```

**Comment by dharmaturtle** [comment]
Great suggestion Tyson!

I also like the idea of removing the primitive types from `AutoGenConfig`, though not necessarily as part of this PR. They wouldn't really be `Overrides` in that case... more like `Registrations`. Perhaps a rename is in order in preparation for that PR?

**Comment by cmeeren** [comment]
> I also like the idea of removing the primitive types from `AutoGenConfig`, though not necessarily as part of this PR.

I can do another PR for this.

> They wouldn't really be `Overrides` in that case... more like `Registrations`. Perhaps a rename is in order in preparation for that PR?

Yes, as I mentioned above, perhaps `Generators` would be a good name.

**Comment by cmeeren** [comment]
In any case, are people happy with this PR? Can I merge it?

**Comment by dharmaturtle** [comment]
>Yes, as I mentioned above, perhaps Generators would be a good name.

Ah whoops, I missed reading that. I had the same idea so I just skimmed that comment >_>

FWIW I'm happy with this PR!

**Comment by TysonMN** [comment]
> I am talking about removing all the primitive `Byte`, `Int16` etc. members of `AutoGenConfig`...

I understand now.  I didn't know what fields were on [`AutoGenConfig`](https://github.com/hedgehogqa/fsharp-hedgehog-experimental/blob/92c8c60f1aae48459f2604545ab94cc1fa1300a4/src/Hedgehog.Experimental/Gen.fs#L7).

**Comment by TysonMN** [comment]
I also like this PR.  Commit https://github.com/TysonMN/fsharp-hedgehog-experimental/commit/b0055109ba64692bb0e7488a10ff817342d41476 has some small changes to the implementation that think are improvements.

File: src/Hedgehog.Experimental/Gen.fs [hunk 2]
```diff
@@ -382,154 +397,158 @@ module GenX =
...
-        | Some ctor ->
-          ctor.Accept {
-          new IConstructorVisitor<'a, Gen<'a>> with
-            member __.Visit<'CtorParams> (ctor : ShapeConstructor<'a, 'CtorParams>) =
-              let paramGen = autoInner<'CtorParams> config recursionDepths
-              gen { 
-                let! args = paramGen
-                return ctor.Invoke args
+    match config.Overrides.Wrapped.TryFind typeof<'a>.FullName with
+    | Some gen -> gen |> Gen.map unbox<'a> |> wrap
```
Comment (TysonMN) [inline src/Hedgehog.Experimental/Gen.fs:401]: What is `wrap` doing there?  If `'b` is `'a`, is `wrap` just the identity function?  The code still compiles and passes all tests if I remove it.  I am guessing that it is a copy-paste from the other cases below.

Comment (cmeeren) [reply src/Hedgehog.Experimental/Gen.fs:401]: IIRC it's from before I took over the auto-gen code (which still fundamentally works the same was as back then). At the time there was a lot of stuff there I didn't understand, so I just left it. I've never understood `wrap`, but never tried removing it.

**Comment by cmeeren** [comment]
> 
> 
> I also like this PR. Commit [TysonMN@b005510](https://github.com/TysonMN/fsharp-hedgehog-experimental/commit/b0055109ba64692bb0e7488a10ff817342d41476) has some small changes to the implementation that think are improvements.

Sure, why not. If you create a PR to this PR, I'll merge.

**Comment by moodmosaic** [comment]
>I also like this PR.  Commit https://github.com/TysonMN/fsharp-hedgehog-experimental/commit/b0055109ba64692bb0e7488a10ff817342d41476 has some small changes to the implementation that think are improvements.

These are good suggestions :+1:

**Comment by moodmosaic** [comment]
PR looks reasonable to me :+1:

**Comment by cmeeren** [comment]
I went ahead and implemented @TysonMN's changes myself and will merge now. Thanks!

File: src/Hedgehog.Experimental/Gen.fs (hunk 2 above)
Comment (cmeeren) [reply src/Hedgehog.Experimental/Gen.fs:408]: I'm confused by what you mean by

> If `'b` is `'a`

Are they the same?

Also, I tried removing `wrap`, but it won't compile.

Comment (TysonMN) [reply src/Hedgehog.Experimental/Gen.fs:408]: The tooltip says `'b is 'a`, so my "if `'b` is `'a`" was meant to be rhetorical.  Sorry for being unclear.

I will try removing `unwrap` again later.

Comment (TysonMN) [reply src/Hedgehog.Experimental/Gen.fs:408]: Continuing this discussion in PR #46

**Pull request merged**